### DISCUSSION TITLE
Localization, modular and Composable-Architecture-friendly?

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-composable-architecture#1003] [Q&A]
Hello everyone.

Does anyone have a recommendation for a good approach to localization that is both modular and that fits well with the Composable Architecture? I have an inkling that localization is akin to an external API and, therefore, could/should be considered part of the environment.

Many thanks!

**Comment by JacksonUtsch** [comment]
One note, you can include localization files in each module's resources but it seems an app target requires localization in its bundle in order to pull the others. So my "modular" approach is just to separate by module/feature.

**Reply by iampatbrown** [reply]
I normally use this helper when working with localization (and other resources). It also helps get previews working when modules have images/fonts.

```swift 
import Foundation
import SwiftUI

/// https://developer.apple.com/forums/thread/664295
private class BundleFinder {}

public let bundle: Bundle = {
  let bundleName = "FrutaTCA_Localization"
  let bundleResourceURL = Bundle(for: BundleFinder.self).resourceURL
  let candidates = [
    // Bundle should be present here when the package is linked into an App.
    Bundle.main.resourceURL,
    // Bundle should be present here when the package is linked into a framework.
    bundleResourceURL,
    // For command-line tools.
    Bundle.main.bundleURL,
    /* Bundle should be present here when running previews from a different package "…/Debug-iphonesimulator/" */
    bundleResourceURL?.deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent(),
    bundleResourceURL?.deletingLastPathComponent().deletingLastPathComponent(),
    bundleResourceURL?.deletingLastPathComponent(),
  ]

  for candidate in candidates {
    let bundlePath = candidate?.appendingPathComponent(bundleName + ".bundle")
    if let bundle = bundlePath.flatMap(Bundle.init(url:)) { return bundle }
  }
  fatalError("unable to find bundle named \(bundleName)")
}()
```

and the target looks like this:

```swift
.target(
  name: "Localization",
  resources: [.process("Resources")]
),
```

![Localization](https://user-images.githubusercontent.com/19797367/155033531-4fa18683-5672-4da3-896c-507ee7aa3627.png)

```swift
Text(self.label, bundle: Localization.bundle)
  .frame(width: 150)
  .padding(.vertical)
```

**Reply by JacksonUtsch** [reply]
Does this do anything more than just getting the module bundle?

```swift
extension Bundle {
	public static let moduleName = Bundle.module
}
```

I previously used a bundle search like you have due to a bug that has since been resolved.

The `Bundle.module` method does seem to require registering via `CTFontManagerRegisterGraphicsFont`

I also keep my localization per Feature to modularize the keyword usage.

**Reply by gohanlon** [reply]
The custom-bundle finding workaround is still necessary for my app. My circumstance:

* A feature module depends on a shared module
* Shared module has and uses bundled resources
* When running from the feature module, SwiftUI Previews consistently crash due to failing to find the resources used by and located within the shared module

(On Xcode 13.3 beta 2, and several recent GM releases.)

**Reply by JacksonUtsch** [reply]
Ah, gotcha. I've gotten out of the habit of using previews some reason. Sounds about right. Thanks.

**Reply by iampatbrown** [reply]
@JacksonUtsch Yeah I'm pretty sure it's still needed for my projects... mainly to prevent previews from crashing like @gohanlon mentioned. 

I still use `CTFontManagerRegisterGraphicsFont` to register fonts as well.
